src/rdkafka.c

```diff
@@ -270,7 +279,463 @@ void rd_kafka_log0 (const rd_kafka_conf_t *conf,
+                }
+        }
+        rd_ts_t now_ms = rd_clock() / 1000;
+        if (md_lifetime_ms <= now_ms) {
+		rd_kafka_set_fatal_error(rk, RD_KAFKA_RESP_ERR__INVALID_ARG,
+			    "Must supply an unexpired token: now=%lli ms, exp=%lli ms",
+                            now_ms, md_lifetime_ms);
+                return -1;
+        }
+        rwlock_wrlock(&rk->rk_oauthbearer->refresh_lock);
+        // rd_free() ends up as a no-op only on initial invocation
+        rd_free(rk->rk_oauthbearer->md_principal_name);
+        rk->rk_oauthbearer->md_principal_name = rd_strdup(md_principal_name);
+        rd_free(rk->rk_oauthbearer->token_value);
+        rk->rk_oauthbearer->token_value = rd_strdup(token_value);
+        rk->rk_oauthbearer->md_lifetime_ms = md_lifetime_ms;
+        // Schedule a refresh 80% through its remaining lifetime
+        rk->rk_oauthbearer->refresh_after_ms = now_ms + 0.8 * (md_lifetime_ms - now_ms);
+        rd_list_destroy(&rk->rk_oauthbearer->extensions);
+        for (i = 0; i + 1 < extension_size; i += 2) {
+                rd_list_add(&rk->rk_oauthbearer->extensions,
+                        rd_strtup_new(extensions[i], extensions[i + 1]));
+        }
+        rk->rk_oauthbearer->errstr[0] = '\0';
+        rwlock_wrunlock(&rk->rk_oauthbearer->refresh_lock);
+        rd_kafka_dbg(rk, SECURITY, "BRKMAIN",
+                "Waking up waiting brokers after setting token");
+        rd_kafka_all_brokers_wakeup(rk, RD_KAFKA_BROKER_STATE_TRY_CONNECT);
+
+        return 0;
+}
+
+void rd_kafka_oauthbearer_set_token_failure(rd_kafka_t *rk,
+                const char *errstr) {
+        rwlock_wrlock(&rk->rk_oauthbearer->refresh_lock);
+        strncpy(rk->rk_oauthbearer->errstr, errstr,
+                sizeof(rk->rk_oauthbearer->errstr));
+        /* Leave any existing token because it may have some life left */
+        /* Schedule a refresh for 10 seconds later */
+        rk->rk_oauthbearer->refresh_after_ms = rd_clock() / 1000 + 10 * 1000;
+        rwlock_wrunlock(&rk->rk_oauthbearer->refresh_lock);
+}
+
+static void parse_unsecured_jws_config_value_for_prefix(char **loc, const char *prefix,
+                const char value_end_char, char **value, char *errstr, size_t errstr_size) {
+        if (*value) {
+                rd_snprintf(errstr, errstr_size,
+                        "Invalid sasl.oauthbearer.config: multiple '%s' entries", prefix);
+                return;
+        }
+        *loc += strlen(prefix);
+        *value = *loc;
+        while (**loc != '\0' && **loc != value_end_char) {
+                ++*loc;
+        }
+        if (**loc == value_end_char) {
+                // end the string and skip the character
+                **loc = '\0';
+                ++*loc;
+        }
+        // return new allocated memory
+        *value = strdup(*value);
+}
+
+/*
+ * Parse Unsecured JWS config, allocates strings that must be freed
+ */
+static int parse_unsecured_jws_config(const char *cfg,
```

Comment:
There's a lot of arguments to this function, it would be cleaner if it took a string input, and populated a (provided) struct with the values.

Reply:
👍 Didn't include `errstr` and `errstr_size` in the struct.